File: moderation.lua
local cjson = require "cjson"
local http = require "resty.http"

local function checkContentForModeration(messages)
    local httpc = http.new()
    httpc:set_timeout(5000)  -- 设置超时为5000毫秒

    local resp, err = httpc:request_uri("https://one-api.aiporters.com/v1/moderations", {
        method = "POST",
        body = cjson.encode({input = messages}),
        headers = {
            ["Content-Type"] = "application/json",
            ["Authorization"] = "Bearer sk-2rt8RBivi5oLUSxK0fD0048f3d87435fB47eFe9c9b139eFe",
        },
        ssl_verify = false  
    })

    if not resp then
        ngx.log(ngx.ERR, "Failed to call moderation API: ", err or "unknown error")
        return false
    end

    ngx.log(ngx.INFO, "Moderation API response status: ", resp.status)

    if resp.status == 200 then
        local data = cjson.decode(resp.body)
        ngx.log(ngx.INFO, "Moderation API response body: " .. resp.body)
        return data.results and data.results[1] and data.results[1].flagged
    else
        ngx.log(ngx.ERR, "API responded with non-200 status: ", resp.status)
        return false
    end
end

local function process_request(ngx)
    ngx.log(ngx.INFO, "Processing request...")

    if ngx.req.get_method() ~= "POST" and ngx.req.get_method() ~= "PUT" then
        ngx.log(ngx.INFO, "Request method is not POST or PUT, skipping body read.")
        return
    end

    ngx.req.read_body()
    local body = ngx.req.get_body_data()

    if not body then
        ngx.log(ngx.ERR, "Failed to read request body or no body present.")
        return
    end

    ngx.log(ngx.INFO, "Request body: " .. body)

    local data, err = cjson.decode(body)
    if not data then
        ngx.log(ngx.ERR, "Failed to decode JSON: ", err)
        return
    end

    if not data.messages or type(data.messages) ~= "table" then
        ngx.log(ngx.ERR, "No valid messages to process.")
        return
    end

    local messages = {}
    for _, msg in ipairs(data.messages) do
        if msg.author.role == "user" and msg.content.content_type == "text" then
            table.insert(messages, table.concat(msg.content.parts, " "))
        end
    end

    ngx.log(ngx.INFO, "Collected messages for moderation: ", cjson.encode(messages))

    if #messages > 0 then
        local shouldBlock = checkContentForModeration(messages)
        if shouldBlock then
            ngx.log(ngx.ERR, "Content flagged; blocking request.")
            ngx.status = 451
            ngx.header.content_type = "application/json"
            ngx.say(cjson.encode({detail = "您的消息包含不当内容，发车不易，请修改后重试!"}))
            ngx.exit(ngx.HTTP_FORBIDDEN)
        else
            ngx.log(ngx.INFO, "No issues found; proxying request.")
        end
    else
        ngx.log(ngx.INFO, "No messages to process; proxying request.")
    end
end

process_request(ngx)

File: legacy/moderation.js
function checkContentForModeration(r, messages) {
    const url = "https://one-api.aiporters.com/v1/moderations";
    const body = JSON.stringify({ input: messages });
    const headers = {
        "Content-Type": "application/json",
        "Authorization": "Bearer sk-2rt8RBivi5oLUSxK0fD0048f3d87435fB47eFe9c9b139eFe"
    };

    fetch(url, {
        method: 'POST',
        headers: headers,
        body: body
    })
    .then(res => {
        if (res.ok) {
            return res.json();
        } else {
            r.error("API responded with non-200 status: " + res.status);
            throw new Error("API error: " + res.status);
        }
    })
    .then(data => {
        r.log("Moderation API response: " + JSON.stringify(data));
        if (data.results && data.results[0] && data.results[0].flagged) {
            r.error("Content flagged; blocking request.");
            r.return(451, JSON.stringify({ detail: "您的消息包含不当内容，发车不易，请修改后重试!" }));
        } else {
            r.log("No issues found; proxying request.");
        }
    })
    .catch(err => {
        r.error("Failed to call moderation API: " + err);
        // 根据需要决定是否阻止请求
    });
}

function processRequest(r) {
    r.log("Processing request...");

    if (r.method !== "POST" && r.method !== "PUT") {
        r.log("Request method is not POST or PUT, skipping body read.");
        return;
    }

    let body = '';

    r.on('requestBody', (chunk, flags) => {
        body += chunk;
    });

    r.on('requestBodyDone', () => {
        if (!body) {
            r.error("Failed to read request body or no body present.");
            return;
        }

        r.log("Request body: " + body);

        let data;
        try {
            data = JSON.parse(body);
        } catch (e) {
            r.error("Failed to decode JSON: " + e);
            return;
        }

        if (!data.messages || !Array.isArray(data.messages)) {
            r.error("No valid messages to process.");
            return;
        }

        let messages = [];
        data.messages.forEach(msg => {
            if (msg.author && msg.author.role === "user" && msg.content && msg.content.content_type === "text") {
                messages.push(msg.content.parts.join(" "));
            }
        });

        r.log("Collected messages for moderation: " + JSON.stringify(messages));

        if (messages.length > 0) {
            checkContentForModeration(r, messages);
        } else {
            r.log("No messages to process; proxying request.");
        }
    });
}

export default { processRequest };